### PR TITLE
feature:S3C-4200 Assume role backbeat API Integration - 1

### PR DESCRIPTION
**Pull request by saitharunthodupunuru**

### CODE REVIEW

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -16,11 +19,13 @@ osis.scality.vault.secret-key=UEEu8tYlsOGGrgf4DAiSZD6apVNPUWqRiPG0nTB6
...
 osis.scality.vault.cache.listAccounts.maxCapacity=1000
 osis.scality.vault.cache.listAccounts.expirationInMS=60000
+osis.scality.vault.cache.assumeRole.maxCapacity=1000
+osis.scality.vault.cache.assumeRole.expirationInMS=60000
 
-# S3 Config
-osis.scality.s3.endpoint=http://scality.osis.ose.vmware.com:31383
-#osis.scality.s3.access-key=094BC4MZSZMET634FXZ0
-#osis.scality.s3.secret-key=Gpyohi7gTe0YriLNhURixitKoZ8IVAsCPrDDZxh6
+# S3 Config - Vault endpoint and S3 port with super admin credentials
```
Comment (anurag4DSB) [inline src/main/resources/application.properties:25]: Vault has `superAdmin` credentials. S3/Cloudserver doesn't understand superAdmin credentials. 
Can be a little confusing, mentioning Vault should be enough if using `superAdmin` Creds

File: vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java [hunk 2]
```diff
@@ -6,50 +6,63 @@
...
+  public VaultAdminImpl(String accessKey, String secretKey, String endpoint, String s3Endpoint) {
     validEndpoint(endpoint);
     this.vaultAccountClient = new AccountServicesClient(
             new BasicAWSCredentials(accessKey, secretKey));
     vaultAccountClient.setEndpoint(endpoint);
 
+
+    this.vaultSTSClient = new SecurityTokenServicesClient(
+            new BasicAWSCredentials(accessKey, secretKey));
+    vaultSTSClient.setEndpoint(s3Endpoint);
```
Comment (anurag4DSB) [inline vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java:64]: Shouldn't it be namedd STS Endpoint.
`<Vault-IP>:8800`

File: vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java [hunk 3]
```diff
@@ -58,16 +71,23 @@ public VaultAdminImpl(String accessKey, String secretKey, String endpoint) {
...
-  public VaultAdminImpl(AccountServicesClient vaultAccountClient, String endpoint) {
+  public VaultAdminImpl(AccountServicesClient vaultAccountClient, SecurityTokenServicesClient vaultSTSClient,
+                        String endpoint, String s3Endpoint) {
     validEndpoint(endpoint);
     this.vaultAccountClient = vaultAccountClient;
     vaultAccountClient.setEndpoint(endpoint);
+
+    validEndpoint(s3Endpoint);
+    this.vaultSTSClient = vaultSTSClient;
+    vaultSTSClient.setEndpoint(s3Endpoint);
```
Comment (anurag4DSB) [inline vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java:82]: Same here. STS endpoint.

File: vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java [hunk 4]
```diff
@@ -42,6 +47,13 @@
 
   protected static final String ENTITY_EXISTS_ERR = "The request was rejected because it attempted to create a resource that already exists.";
 
+  public static final String TEST_ACCESS_KEY = "access_key";
+  public static final String TEST_SECRET_KEY = "secret_key";
+  public static final String TEST_SESSION_TOKEN = "session_token";
+  public static final String TEST_SESSION_NAME = "session";
+  public static final String TEST_ROLE_ARN = "arn:aws:iam::0123456789:role/osis";
```
Comment (anurag4DSB) [inline vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java:54]: An account ID is 12 digits.

File: vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java [hunk 5]
```diff
@@ -42,6 +47,13 @@
...
+  public static final String TEST_ASSUMED_USER_ARN = "arn:aws:sts::0123456789:assumed-role/osis/session1";
```
Comment (anurag4DSB) [inline vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java:55]: Same here. You can use something like `123123123123`
12 digits.

File: vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java [hunk 6]
```diff
@@ -50,11 +62,13 @@ public void init() throws IOException {
...
   protected void initMocks() {
     accountServicesClient = mock(AccountServicesClient.class);
+    stsClient = mock(SecurityTokenServicesClient.class);
 
     initCreateAccountMocks();
     initListAccountsMocks();
+    initAssumeRoleMocks();
 
-    vaultAdminImpl = new VaultAdminImpl(accountServicesClient, adminEndpoint);
+    vaultAdminImpl = new VaultAdminImpl(accountServicesClient, stsClient, adminEndpoint, s3Endpoint);
```
Comment (anurag4DSB) [inline vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java:71]: Should it be VaultEndpoint?
S3 and vault Endpoint will/can be different(port numbers).
Making it future proof now, will help.

File: vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/VaultAdminImplTest.java [hunk 7]
```diff
@@ -208,7 +211,7 @@ public void listAccounts500ServiceResponse() {
     @Test
     public void testVaultAdminImplInvalidEndpoint() {
         assertThrows(IllegalArgumentException.class,
-                () -> new VaultAdminImpl(accountServicesClient, "dummy_endpoint"),
+                () -> new VaultAdminImpl(accountServicesClient, stsClient, "dummy_endpoint", s3Endpoint),
```
Comment (anurag4DSB) [inline vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/VaultAdminImplTest.java:214]: Same here. May be Vault endpoint.

**Review by anurag4DSB — CHANGES_REQUESTED** [review]

File: src/main/resources/application.properties (hunk 1 above)
Comment (saitharunthodupunuru) [reply src/main/resources/application.properties:25]: Ok

File: vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java (hunk 2 above)
Comment (saitharunthodupunuru) [reply vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java:64]: No `AssumeRoleBackbeat` is called with s3 port 8500 so s3 URl

Comment (saitharunthodupunuru) [reply vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java:64]: Please let me know if we can invoke route `AssumeRoleBackbeat` with port 8800

Comment (anurag4DSB) [reply vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java:64]: Are you making a call to S3 ie. clousderver and then asking s3 to make a call to vault on port 8500?
If yes, then good. If OSIS is making a call directly on port 8500, then its vault.

Yes, you are right, its port 8500.

File: src/main/resources/application.properties (hunk 1 above)
Comment (saitharunthodupunuru) [reply src/main/resources/application.properties]: Done in 9101950281bb1e469ee5c3cbe333dcaa4c9a3953

File: vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java (hunk 4 above)
Comment (saitharunthodupunuru) [reply vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java]: Done in 9101950281bb1e469ee5c3cbe333dcaa4c9a3953

File: vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java (hunk 6 above)
Comment (saitharunthodupunuru) [reply vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java]: Done in 9101950281bb1e469ee5c3cbe333dcaa4c9a3953

File: vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java (hunk 5 above)
Comment (saitharunthodupunuru) [reply vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/BaseTest.java]: Done in 9101950281bb1e469ee5c3cbe333dcaa4c9a3953

File: vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/VaultAdminImplTest.java (hunk 7 above)
Comment (saitharunthodupunuru) [reply vault-admin-client/src/test/java/com/scality/osis/vaultadmin/impl/VaultAdminImplTest.java]: Done in 9101950281bb1e469ee5c3cbe333dcaa4c9a3953

File: src/main/resources/application.properties [hunk 8]
```diff
@@ -22,10 +22,8 @@ osis.scality.vault.cache.listAccounts.expirationInMS=60000
 osis.scality.vault.cache.assumeRole.maxCapacity=1000
 osis.scality.vault.cache.assumeRole.expirationInMS=60000
 
-# S3 Config - Vault endpoint and S3 port with super admin credentials
+# S3 Config - Vault endpoint and S3 port
 osis.scality.s3.endpoint=http://localhost:8500
```
Comment (anurag4DSB) [inline src/main/resources/application.properties:26]: Not sure why I cannot suggest but here we go:

```
# Vault S3 interface configuration
osis.scality.vault-internal-s3-interface.endpoint=http://localhost:8500
#osis.scality.admin.access-key=D4IT2AWSB588GO5J9T00
#osis.scality.admin.secret-key=UEEu8tYlsOGGrgf4DAiSZD6apVNPUWqRiPG0nTB6
```
Feel free to change the case of `vault-internal-s3--interface`. For Example: camel case `caultInternalS3Interface`


This is not S3 endpoint. Its an S3 Interface but vault service. Also, the keys are admin keys and not s3 keys. S3/Cloudserver doesn't understand admin routes.

S3 Endpoint port number is 8000 served by cloudserver.

**Review by anurag4DSB — CHANGES_REQUESTED** [review]
I see the confusion.

Vault serves on 3 ports:
`S3` Interface: 8500 - used by cloudserver, backbeat, internal services for calls like Authv4, Authv2, AssumeRoleBackbeat
Reference:
* https://github.com/scality/Vault/blob/development/7.10/config.json#L10
* https://github.com/scality/Vault/blob/development/7.10/lib/routes.js#L2

`administration` interface: 8600 - used for add admin calls(accounts) and IAM calls
Reference:
* https://github.com/scality/Vault/blob/development/7.10/config.json#L18
* https://github.com/scality/Vault/blob/development/7.10/lib/routes.js#L107

`sts` interface: 8800 - used for STS calls
Reference: 
* https://github.com/scality/Vault/blob/development/7.10/config.json#L22
* https://github.com/scality/Vault/blob/development/7.10/lib/routes.js#L94


Hence we need to name the variables and config in OSIS appropiately.

**Review by ilkescality — APPROVED** [review]

File: vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java (hunk 2 above)
Comment (saitharunthodupunuru) [reply vault-admin-client/src/main/java/com/scality/osis/vaultadmin/impl/VaultAdminImpl.java]: fixed it in 916eec1d0143504a70d18803cb68a1a3643cabaa

File: src/main/resources/application.properties (hunk 8 above)
Comment (saitharunthodupunuru) [reply src/main/resources/application.properties]: done in 916eec1d0143504a70d18803cb68a1a3643cabaa

**Review by anurag4DSB — APPROVED** [review]

**Comment by saitharunthodupunuru** [comment]
Squashed the review commits

**Pull request merged**